Extensions: Add extensions building

sirreal opened the pull request:
- Adds block building to Jetpack — `yarn build-extensions`
- Update tooling (pre-commit, eslint, reformat-files) for extensions

More fun? Watch and build! `yarn build-extensions --watch` 💥 

## Blockers
- [x] #11677 (fix CommonJS / esmodules imports)
- [x] #10810 (Upgrade babel for existing React app: Jetpack dashboard/settings)

## Testing

- `yarn` (update/install deps)
- Extensions build as expected and work correctly:
  - Clean (`yarn clean-client`) -- nothing in `_inc/blocks`
  - Build (`yarn build-extensions`) Built and working!
- Check a few things in the output:
  - Output is what we expect:
    - no references to React or ReactDOM
    - es6 is properly transpiled (e.g. no plain `class` declarations)
  - Production builds are minified (`NODE_ENV=production yarn build-extensions`)
- Run a full build and check for regressions (`yarn clean-client && yarn build`)
  - Do the blocks and extensions continue to work well?
  - React-based Dashboard works well?

## Changelog

- Bring the Block Editor extensions source and build tooling into the Jetpack repo

## Concerns

After #10810 lands, none 😎 

~Two versions of babel and babel deps~

~The React based Jetpack dashboard depends on older versions of `babel`. I suspect it's find to use both because these versions happen not to collide (`babel-thing@6` vs `@babel/thing@7`)~

sirreal committed:
Add build system for block editor extensions

Up to date as of Calypso@e4862897aec6ccd7e3ef7120df19f20df216a07a:

https://github.com/Automattic/wp-calypso/commit/e4862897aec6ccd7e3ef7120df19f20df216a07a

sirreal commented on package.json:
```diff
@@ -1,169 +1,203 @@
+	},
+	"devDependencies": {
+		"@automattic/calypso-build": "1.0.0-alpha.0",
+		"@automattic/calypso-color-schemes": "1.0.0-alpha.0",
```
This version is broken. We need `1.0.0-alpha.1` when https://github.com/Automattic/wp-calypso/pull/31628 lands and is published.

```suggestion
		"@automattic/calypso-color-schemes": "1.0.0-alpha.1",
```

simison commented on babel.config.blocks.js:
```diff
@@ -0,0 +1,60 @@
+		[
+			'@babel/env',
+			{
+				targets: { browsers: [ 'last 2 versions', 'Safari >= 10', 'iOS >= 10', 'ie >= 11' ] },
```
Let's bring this change over:

https://github.com/Automattic/wp-calypso/pull/31464/files#diff-0e759b21864de1336af9c450cff5ad20R9

simison replied:
Actually let's just leave this undefined so that Babel picks up these from `package.json`?

sirreal replied:
Did it 👍

simison commented:
Should we attempt already here to combine this with Jetpack's babel config:

https://github.com/Automattic/jetpack/blob/2d4fcb85393ec11580e3e8609d5018b66fcfec3e/.babelrc

...by using overrides?

We'll need to merge these eventually anyway, I suppose.

simison commented on package.json:
```diff
@@ -1,169 +1,203 @@
-    "node": ">=8",
-    "yarn": "^1.3.2"
-  }
+	"name": "Jetpack",
```
This whitespace change feels like it could be in a separate PR. ;-)

simison replied:
PRed whitespace change here: https://github.com/Automattic/jetpack/pull/11646

sirreal commented:
> Should we attempt already here to combine this with Jetpack's babel config:

I like keeping these separate for now, but I'd love to see the same build produce them in the future.

I'm actually feeling quite good about landing this even before 7.2, but I don't want to overhaul the Dashboard build in this small window. 7.3 would be a good target for that.

simison commented:
>The React based Jetpack dashboard depends on older versions of babel. I suspect it's find to use both because these versions happen not to collide (babel-thing@6 vs @babel/thing@7)

PR for Babel updates: https://github.com/Automattic/jetpack/pull/10810

simison commented:
I nudged the Babel upgrade PR https://github.com/Automattic/jetpack/pull/10810 forward little bit — looks like it could be finished after Prettier changes are merged https://github.com/Automattic/jetpack/pull/11507

lezama commented on babel.config.blocks.js:
```diff
@@ -3,7 +3,6 @@ const config = {
 		[
 			'@babel/env',
 			{
-				targets: { browsers: [ 'last 2 versions', 'Safari >= 10', 'iOS >= 10', 'ie >= 11' ] },
```
why we don't need this?

sirreal replied:
It uses the browserlist defined elsewhere:

https://github.com/Automattic/jetpack/blob/808170fbe623b720fea37d04a377137b6b448c55/package.json#L13-L15

simison replied:
It's basically duplicate of the browserslist config here:

https://github.com/Automattic/jetpack/blob/48bd35a11c690c8bd8abf398ac5e324ad1d708c8/package.json#L13-L15

The same config got removed from Calypso Babel just recently as well, for the same reason: https://github.com/Automattic/wp-calypso/pull/31561

As a result browser compatibility requirement are defined in one place only. `:success:`

lezama commented:
> I'm actually feeling quite good about landing this even before 7.2, but I don't want to overhaul the Dashboard build in this small window. 7.3 would be a good target for that.

I would do it for 7.2, it would hit the entire test cycle. Not sure it's going to get more tested than that if we do it for 7.3. it will also free us from this task in the future :)

sirreal commented:
> > I'm actually feeling quite good about landing this even before 7.2, but I don't want to overhaul the Dashboard build in this small window. 7.3 would be a good target for that.
> 
> I would do it for 7.2, it would hit the entire test cycle. Not sure it's going to get more tested than that if we do it for 7.3. it will also free us from this task in the future :)

To clarify my original comment: I'd land this _as is_ for 7.2, and work on a more unified build for future releases.

simison commented:
It'll get more testing first in `master` and then during 7.2 Beta week if we combine these files only after 7.2 is branched next week.

Babel upgrade PR https://github.com/Automattic/jetpack/pull/10810 isn't quite ready yet it seems so merging two separate configs at first will get us forward and not blocked with Babel upgrade. It's a good mid-step towards one config, maybe just shared from Calypso build package eventually?

I agree with sentiment and excitement of feeling like getting to a place where we want to be earlier is tempting. :-)

simison commented:
Rebased.

ockham commented on .eslintrc.js:
```diff
@@ -95,5 +98,34 @@ module.exports = {
+				"wpcalypso/jsx-classname-namespace": 0,
+			},
+		},
+	]
```
Hmm, so we didn't include `jetpack-blocks`' [`.eslintrc.js`](https://github.com/Automattic/wp-calypso/blob/24daf03893351349cc59100310f85173d867100a/packages/jetpack-blocks/.eslintrc.js) in the move? :slightly_frowning_face:

simison replied:
Rules from that config are fairly small but it seems like because this PR was originally against another branch than `master`, CI linter didn't run at all so we didn't catch other linter issues here.

simison replied:
I brought changes over from jetpack-blocks' .eslintrc.js in https://github.com/Automattic/jetpack/pull/11640/commits/da2c4515aa6dc708a1170ad46b542ea77f344cb1

sirreal replied:
No, the lint errors didn't pop up and we didn't pull it in. Ideally we share published prettier and eslint from Calypso, but we're not quite there. I'd like this entire eslintconfig to be something like `extends: [ 'wp-calypso/react', 'wp-calypso/gutenberg' ]`.

ockham commented:
This currently gives me an `_inc/blocks` weighing 4.6 megs (the :lion:'s share being `editor.js` and `editor-beta.js` at about 2 megs each). By comparison, WP.com's current `_inc/blocks-staging` is 3.3 megs (`editor.js` and `editor-beta.js` are 1.3 each). I wonder what's bloating us here...

sirreal commented:
@ockham are you running `NODE_ENV=production yarn build-extensions`? I get `2.4MB`

Here's what I see:

```
& yarn clean-clean
$ NODE_ENV=production yarn build-extensions
$ du -ah _inc/blocks

8.0K    _inc/blocks/tiled-gallery/view.rtl.css
 16K    _inc/blocks/tiled-gallery/view.js
8.0K    _inc/blocks/tiled-gallery/view.css
 32K    _inc/blocks/tiled-gallery
 20K    _inc/blocks/vendors~swiper.css
 60K    _inc/blocks/editor.css
 60K    _inc/blocks/editor.rtl.css
 84K    _inc/blocks/images/map-theme_black_and_white-1ead5946ca104d83676d6e3410e1d733.jpg
8.0K    _inc/blocks/images/paypal-button-2x-fe4d34770a47484f401cecbb892f8456.png
4.0K    _inc/blocks/images/oval-3cc7669d571aef4e12f34b349e42d390.svg
112K    _inc/blocks/images/map-theme_default-2ceb449b599dbcbe2a90fead5a5f3824.jpg
108K    _inc/blocks/images/map-theme_terrain-2b6e6c1c8d09cbdc58a4c0653be1a6e3.jpg
8.0K    _inc/blocks/images/paypal-button-1e53882e702881f8dfd958c141e65383.png
160K    _inc/blocks/images/map-theme_satellite-c74dc129bda9502fb0fb362bb627577e.jpg
484K    _inc/blocks/images
 24K    _inc/blocks/vendors~map/mapbox-gl.rtl.css
652K    _inc/blocks/vendors~map/mapbox-gl.js
 24K    _inc/blocks/vendors~map/mapbox-gl.css
700K    _inc/blocks/vendors~map
 20K    _inc/blocks/vendors~swiper.rtl.css
4.0K    _inc/blocks/repeat-visitor/view.js
4.0K    _inc/blocks/repeat-visitor
344K    _inc/blocks/editor-beta.js
4.0K    _inc/blocks/mailchimp/view.rtl.css
4.0K    _inc/blocks/mailchimp/view.js
4.0K    _inc/blocks/mailchimp/view.css
 12K    _inc/blocks/mailchimp
336K    _inc/blocks/editor.js
4.0K    _inc/blocks/map/view.rtl.css
 20K    _inc/blocks/map/view.js
4.0K    _inc/blocks/map/view.css
 28K    _inc/blocks/map
128K    _inc/blocks/vendors~swiper.js
 60K    _inc/blocks/editor-beta.css
4.0K    _inc/blocks/contact-info/view.rtl.css
4.0K    _inc/blocks/contact-info/view.js
4.0K    _inc/blocks/contact-info/view.css
 12K    _inc/blocks/contact-info
4.0K    _inc/blocks/index.json
8.0K    _inc/blocks/slideshow/view.rtl.css
 16K    _inc/blocks/slideshow/view.js
8.0K    _inc/blocks/slideshow/view.css
 32K    _inc/blocks/slideshow
 60K    _inc/blocks/editor-beta.rtl.css
4.0K    _inc/blocks/gif/view.rtl.css
4.0K    _inc/blocks/gif/view.js
4.0K    _inc/blocks/gif/view.css
 12K    _inc/blocks/gif
2.4M    _inc/blocks
```

simison commented:
Let's test if we can build a working bundle for wpcom with this setup because of these `~` symbols:
```
 24K    _inc/blocks/vendors~map/mapbox-gl.rtl.css
652K    _inc/blocks/vendors~map/mapbox-gl.js
 24K    _inc/blocks/vendors~map/mapbox-gl.css
700K    _inc/blocks/vendors~map
 20K    _inc/blocks/vendors~swiper.rtl.css
```

We had issues with `@` symbol previously https://github.com/Automattic/wp-calypso/issues/29284 but we assume it was because of CircleCI and sync scripts.

sirreal commented:
> we assume it was because of CircleCI and sync scripts.

I tracked it down and am confident it was our home grown sync scripts. I've tested the map and see requests like `…/wp-content/plugins/jetpack/_inc/blocks/vendors~map/mapbox-gl.js` with no issues 👍

ockham commented:
Ah, my bad :roll_eyes: Forgot about setting `NODE_ENV` :man_facepalming:

sirreal commented:
I've built up a branch with #10810 and #11677: [__extensions-build-plus-babelupgrade](https://github.com/Automattic/jetpack/tree/__extensions-build-plus-babelupgrade)

The Jetpack dashboard/settings as well the Jetpack extensions in the block editor all seem to build and function well 🎉

simison commented on package.json:
```diff
@@ -126,8 +149,11 @@
     "jshint-stylish": "2.2.1",
     "json-loader": "0.5.7",
     "lodash": "4.17.11",
+    "mapbox-gl": "0.53.1",
```
Just noting that we'll need to clean at least this dependency out from Calypso root `package.json`:

https://github.com/Automattic/wp-calypso/blob/e4862897aec6ccd7e3ef7120df19f20df216a07a/package.json#L143

I expected to find it in Jetpack blocks' `package.json` instead so that's why I'm flagging it here: https://github.com/Automattic/wp-calypso/blob/612b387384a9a85b7c45238650af7c0ef2e724c3/packages/jetpack-blocks/package.json

sirreal replied:
Should pull that into https://github.com/Automattic/wp-calypso/pull/31759

I believe also the `markdown-it` dep 👍

simison committed:
Move "repeat visitor" block to production from beta-list

...to match with Jetpack 7.2 release of Blocks with `@automattic/jetpack-blocks@15.0.0`.

ockham commented:
Rebasing now...

ockham commented:
Needs some more cleanup/syncing of Babel config, I'll take care of that.

ockham committed:
babel.config.extensions.js: Sync with Calypso

ockham commented on babel.config.extensions.js:
```diff
@@ -0,0 +1,58 @@
+	],
+	overrides: [
+		{
+			test: './src',
```
Hmm, this goes back to this code's Calypso days. No `src/` directory in `extensions/`. We might just merge the override into the top level `plugins`.

ockham replied:
875f5f2

ockham committed:
Bump some Babel versions

kraftbj commented on babel.config.extensions.js:
```diff
@@ -0,0 +1,58 @@
+			'@babel/env',
+			{
+				useBuiltIns: 'entry',
+				corejs: 2,
```
this is, I think, what is throwing the error Travis saw.

ockham replied:
I just bumped some Babel package versions to fix this :crossed_fingers:

ockham committed:
Remove obsolete babel-register

ockham committed:
Add missing dependencies

ockham committed:
Merge overrides into top-level babel config

ockham committed:
Drop an obsolete dep

ockham commented:
Okay, this should be ready for another look :+1:

simison commented:
When running `yarn watch`:
- `_inc/blocks`: missing 
- `_inc/build`: works

When running `yarn build`, both get generated.

simison commented:
When building blocks in this branch, I'm consistently getting `TypeError: "this.swiperInstance is undefined"` for the Slideshow block.

`/wp-content/plugins/jetpack/_inc/blocks/vendors~swiper.js` does load file in networks tab so I'm not sure what could be the issue.

ockham commented:
> When building blocks in this branch, I'm consistently getting `TypeError: "this.swiperInstance is undefined"` for the Slideshow block.
> 
> `/wp-content/plugins/jetpack/_inc/blocks/vendors~swiper.js` does load file in networks tab so I'm not sure what could be the issue.

I can't seem to repro. Are you saying you're getting that error while building the blocks, i.e. in your terminal upon `yarn build-extensions`?

sirreal commented on .svnignore:
```diff
@@ -1,3 +1,4 @@
+.babelrc*
```
Can drop this now after #10810 

```suggestion
```

ockham replied:
Oops, missed that one. My bad.

ockham replied:
#11697 11697

simison commented:
>I can't seem to repro. Are you saying you're getting that error while building the blocks, i.e. in your terminal upon yarn build-extensions?

Confirmed that this is happening outside this PR in `master` as well, when using Gutenberg 3.5 plugin (and not the latest from Github).

p1553628166525700-slack-jetpack-gutenberg

simison reviewed (approved):
Works well; we just need to follow up to add this to `yarn watch` command as well.

ockham commented:
Thanks all! Going to land this.
